torchvision/prototype/datasets/utils/_resource.py

```diff
@@ -141,8 +143,40 @@ def __init__(
         super().__init__(file_name=file_name or pathlib.Path(urlparse(url).path).name, **kwargs)
         self.url = url
         self.mirrors = mirrors
+        self._resolved = False
+
+    @property
+    def resolved(self) -> bool:
+        return self._resolved
+
+    def resolve(self) -> Optional[OnlineResource]:
+        redirect_url = _get_redirect_url(self.url)
+        if redirect_url == self.url:
+            self._resolved = True
+            return self
+
+        meta = {
+            attr.lstrip("_"): getattr(self, attr)
+            for attr in (
+                "file_name",
+                "sha256",
+                "_preprocess",
+                "_loader",
+            )
+        }
+
+        gdrive_id = _get_google_drive_file_id(redirect_url)
+        if gdrive_id:
+            return GDriveResource(gdrive_id, **meta)
+
+        http_resource = HttpResource(redirect_url, **meta)
+        http_resource._resolved = True
+        return http_resource
 
     def _download(self, root: pathlib.Path) -> None:
+        if not self.resolved:
+            return cast(OnlineResource, self.resolve())._download(root)
+
         for url in itertools.chain((self.url,), self.mirrors or ()):
             try:
                 download_url(url, str(root), filename=self.file_name, md5=None)
```

Comment:
In your first comment you wrote

> This is far from perfect, since the mirrors do not get resolved

Could we instead do the redirection here and handle the mirrors?

Reply:
We could, but that would make stuff more complicated. Given that the plan is to refactor this anyway with `iopatch` and we currently only have a single datasets that has official mirrors, I wouldn't sweat it at the moment.